include/cadmium/basic_model/partial_accumulator.hpp

```diff
@@ -0,0 +1,149 @@
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#ifndef CADMIUM_PARTIAL_ACCUMULATOR_HPP
+#define CADMIUM_PARTIAL_ACCUMULATOR_HPP
+#include<cadmium/modeling/ports.hpp>
+#include<cadmium/modeling/message_bag.hpp>
+#include<limits>
+
+
+namespace cadmium {
+    namespace basic_models {
+
+    /**
+     * @brief PartialAccumulator PDEVS Model.
+     *
+     * PartialAccumulator PDEVS Model:
+     * - X = Numeric
+     * - Y = Numeric
+     * - In_Ports: add<Numeric>, reset, partial
+     * - Out_Ports: outport
+     * - S = {Numeric:total, bool:reseted, int:partial}
+     * - internal({total, true, i}) = {0, false, 0}
+     *   internal({total, false, 1}) = {total, false, 0}
+     * - external({total, b, i}, t, x.add) = {total+x, b, i}
+     *   external({total, b, i}, t, x.reset) = {total, true, i}
+     *   external({total, b, i}, t, x.partial) = {total, b, 1}
+     * - confluence({total, b, i}, 0, x) = external(internal({total, b, i}), 0, x)
+     * - output ({total, b, 1}) = outport:{total}
+     * - time_advance({total, b, 1}) = 0
+     *   time_advance({total, true, i}) = 0
+     *   time_advance ({total, false, 0}) = infinite
+    */
+
+    template<typename VALUE, typename SET> //value is the type of accumulated values
+      struct partial_accumulator_defs {
+        struct sum : public out_port<VALUE> {
+        };
+        struct add : public in_port<VALUE> {
+        };
+        struct reset : public in_port<SET> {
+        };
+        struct partial : public in_port<SET>{
+        };
+    };   
+
+    template<typename VALUE, typename SET, typename TIME> //value is the type of accumulated values
+      struct partial_accumulator {
+        using on_reset=bool;
+        using on_partial=int;

```

Comment:
Maybe you want to define the state as :
template<typename VALUE>
struct state_type{ 
    bool on_reset;
    bool on_partial;
    VALUE accumulated;
};